### PR TITLE
Ryu: add support for maximum significant digits

### LINKED ISSUE #33463: Floating point printing bug whack-a-mole
See https://github.com/JuliaLang/julia/issues/33185#issuecomment-538038803:

> The value "0.025621074" is being printed as "0.025621" and not "0.0256211" in the nightly.

### PR DESCRIPTION
**Pull request by simonbyrne**
This moves the compact truncation to the main reduction stage, which avoids problems with double rounding.

Fixes #33463

### CODE REVIEW

File: test/ryu.jl [hunk 1]
```diff
@@ -748,14 +748,16 @@ end # exp
     @test writecompact(123.4567) == "123.457"
     @test writecompact(0.001234567) == "0.00123457"
     @test writecompact(0.1234567) == "0.123457"
-    @test writecompact(1234567.0) == "123457.0"
+    @test writecompact(1234567.0) == "1.23457e6"
     @test writecompact(12345678910.0) == "1.23457e10"
-    @test writecompact(12345678.0) == "123457.0"
+    @test writecompact(12345678.0) == "1.23457e7"
```
Comment (simonbyrne) [inline test/ryu.jl]: These test cases were incorrect.

**Review by quinnj — APPROVED** [review]
Looks great! Thanks for doing this!

**Comment by simonbyrne** [comment]
The only remaining question is what to do about Float16: the easiest thing I can think of is to change it so that it switches to exponent notation for `1e3` or greater (`eps(Float64(1000)) == 0.5`, `eps(Float16(2048)) == 2)`.

**Comment by quinnj** [comment]
I think that's a good idea; I feel like the previous rules were weirdly complex/inconsistent.

**Comment by simonbyrne** [comment]
Okay, this now gives
```
julia> Float16(999)
Float16(999.0)

julia> Float16(1000)
Float16(1.0e3)
```

**Comment by simonbyrne** [comment]
There are a few weird edge cases where we don't actually print the closest value in order to get a trailing 0 digit:
```
julia> print(nextfloat(0.0,1))
5.0e-324
julia> @printf "%.1e" nextfloat(0.0,1)
4.9e-324
julia> print(nextfloat(0.0,2))
1.0e-323
julia> @printf "%.1e" nextfloat(0.0,2)
9.9e-324
```
and for `Float32`:
```
julia> print(nextfloat(0f0,1))
1.0f-45
julia> @printf "%.1e" nextfloat(0f0,1)
1.4e-45
julia> print(nextfloat(0f0,2))
3.0f-45
julia> @printf "%.1e" nextfloat(0f0,2)
2.8e-45
julia> print(nextfloat(0f0,3))
4.0f-45
julia> @printf "%.1e" nextfloat(0f0,3)
4.2e-45
julia> print(nextfloat(0f0,4))
6.0f-45
julia> @printf "%.1e" nextfloat(0f0,4)
5.6e-45
```

The options:

1. leave as is
2. print the extra digit correctly
3. print only one digit and no decimal, e.g. `5e-324` (since these all have exponents, there is no risk of them being parsed as integers).
  - this appears to be what Python does

File: test/ryu.jl [hunk 2]
```diff
@@ -731,7 +731,7 @@ end
     @test Ryu.writeexp(1e-63, 0) == "1e-63"
     @test Ryu.writeexp(1e-63, 1) == "1.0e-63"
     @test Ryu.writeexp(1e+83, 0) == "1e+83"
-    @test Ryu.writeexp(1e+83, 1) == "1.0e+83"
+324    @test Ryu.writeexp(1e+83, 1) == "1.0e+83"
```
Comment (KristofferC) [inline test/ryu.jl]: 🐛

**Comment by simonbyrne** [comment]
Alright, I've gone for option 3 above, as I think that is the most consistent. Since this and the `Float16` printing are changes from earlier versions I'll leave it for triage to decide.

**Comment by simonbyrne** [comment]
Okay, it has become clear that would be a bigger change than I thought. I'll save that for a new issue.

**Comment by StefanKarpinski** [comment]
That decision seems fine to me 👍

**Comment by quinnj** [comment]
Thanks for doing this @simonbyrne; it makes me feel much better knowing someone else has done a deep dive into this code. :)

**Pull request merged**